### PR TITLE
response structs defined. error handling of invalid input

### PR DESCRIPTION
**Pull request by GeorgeLuo**
adding logic to return start time, status of completion, output of process, end time if completed and exit code if completed for /status endpoint.

/stop and /start endpoint fleshed with errors, as well as generic errors.

added a models.go file to contain responses and requests structs, basic implementations of response structures.

added Command wrapper struct to hold context of process execution.

### CODE REVIEW

File: execUtil.go [hunk 1]
```diff
@@ -3,60 +3,105 @@ package main
...
 
+// TODO prune finished tasks when some max map size is reached
 
 // maps task_id to cmd objects
-var TaskIdCmdMap = make(map[string]*exec.Cmd)
+var TaskIdCommandMap = make(map[string]*CommandWrapper)
 
-// TODO define Command class
+// used for task_id
+var Hostname, err = os.Hostname()
```
Comment (a-palchikov) [inline execUtil.go]: Since this is happening on process init, the expected way to handle initialization is to panic upon error (unless you don't care about the value):
```go
var hostname string
func init() {
	var err error
	hostname, err = os.Hostname()
	if err != nil {
		panic("unable to query hostname")
	}
}
```
or wrapping it up into a function that panics on error:
```go
var hostname = getHostname()
```

File: execUtil.go [hunk 2]
```diff
@@ -3,60 +3,105 @@ package main
...
 
-func GetProcessStatus(task_id string) int {
-  log.Printf("GetProcessStatus task_id=%s, pid=%d, ProcessState=%d", task_id, TaskIdCmdMap[task_id].Process.Pid, TaskIdCmdMap[task_id].ProcessState)
-  return TaskIdCmdMap[task_id].ProcessState.ExitCode()
+type CommandWrapper struct {
+	Command   *exec.Cmd // underlying command
+	Finished  bool // set upon process finish
+  StartTime time.Time
```
Comment (a-palchikov) [inline execUtil.go]: nit: indent is off - I believe the files need to go through `go fmt`

File: execUtil.go [hunk 3]
```diff
@@ -3,60 +3,105 @@ package main
...
+var TaskIdCommandMap = make(map[string]*CommandWrapper)
 
-// TODO define Command class
+// used for task_id
+var Hostname, err = os.Hostname()
 
-func GetProcessStatus(task_id string) int {
-  log.Printf("GetProcessStatus task_id=%s, pid=%d, ProcessState=%d", task_id, TaskIdCmdMap[task_id].Process.Pid, TaskIdCmdMap[task_id].ProcessState)
-  return TaskIdCmdMap[task_id].ProcessState.ExitCode()
+type CommandWrapper struct {
```
Comment (a-palchikov) [inline execUtil.go]: Document the exported types (including the exported attributes) according to the [style guide](https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences).

File: execUtil.go [hunk 4]
```diff
@@ -3,60 +3,105 @@ package main
...
+	Command   *exec.Cmd // underlying command
+	Finished  bool // set upon process finish
+  StartTime time.Time
+  EndTime   time.Time
+  Output    *bytes.Buffer
+}
+
+func GetProcessStatus(task_id string) StatusResponse {
```
Comment (a-palchikov) [inline execUtil.go]: Document the exported methods.
Also using `task_id` as input into internal state without validation is not optimal in the exported method - this should check whether the specified task ID exists in state (and fail appropriately if not) prior to populating the response.

File: execUtil.go [hunk 5]
```diff
@@ -3,60 +3,105 @@ package main
...
-      log.Printf("pre-wait ProcessState=%v", cmd.ProcessState.ExitCode())
       err = cmd.Wait()
-      log.Printf("Command finished with error: %v", err)
-
-      // TODO mark cmd as finished in subroutine
+      TaskIdCommandMap[task_id].Finished = true
+      TaskIdCommandMap[task_id].EndTime = time.Now()
+      if err != nil  {
+        // TODO handle error
+        s.Error = err.Error()
```
Comment (a-palchikov) [inline execUtil.go]: This is not goroutine-safe as the `Error` attribute is potentially accessed from two places - here and the exported `GetProcessStatus` above.
The same applies to `TaskIdCommandMap` - there's no reason this should be a global variable and can be encapsulated into a type with proper access semantics.

File: execUtil.go [hunk 6]
```diff
@@ -3,60 +3,105 @@ package main
...
+  TaskIdCommandMap[task_id] = &CommandWrapper{cmd,false,time.Now(),time.Time{}, &outBuf}
+
+  // async subroutine
   go func() {
-      log.Printf("pre-wait ProcessState=%v", cmd.ProcessState.ExitCode())
       err = cmd.Wait()
-      log.Printf("Command finished with error: %v", err)
-
-      // TODO mark cmd as finished in subroutine
+      TaskIdCommandMap[task_id].Finished = true
```
Comment (a-palchikov) [inline execUtil.go]: You don't need this flag since `EndTime` already serves as an implicit completion indicator.

File: execUtil.go [hunk 7]
```diff
@@ -3,60 +3,105 @@ package main
...
 
-  // check if already finished
+  var s StopResponse
+  s.Task_id = task_id
 
-  err := TaskIdCmdMap[task_id].Process.Kill()
-  if err != nil {
-     log.Fatal(err)
+  // check if task_id in map
+  if commandWrapper, ok := TaskIdCommandMap[task_id]; ok {
```
Comment (a-palchikov) [inline execUtil.go]: Since the `task exists` branch is more involved than the `task does not exist` one, handled the shorter condition first to keep the code left-aligned:
```go
if commandWrapper, ok := TaskIdCommandMap[task_id]; !ok {
	// task_id invalid  <- comment is redundant as it is obvious from the error message
	s.Error = "invalid task_id"
	return s
}
// Continue handling the existing task
```

File: execUtil.go [hunk 8]
```diff
@@ -3,60 +3,105 @@ package main
...
+    //do something here
```
Comment (a-palchikov) [inline execUtil.go]: nit: omit comments that do not explain anything

File: execUtil.go [hunk 9]
```diff
@@ -3,60 +3,105 @@ package main
...
+    err := commandWrapper.Command.Process.Kill()
```
Comment (a-palchikov) [inline execUtil.go]: This is not a graceful way to shut down the process - you should let it close gracefully by sending the appropriate signal before resorting to the `kill` signal.
Also, this only terminates the process itself and not sub-processes it has started. Do you have an idea how to deal with sub-processes?

File: execUtil.go [hunk 10]
```diff
@@ -3,60 +3,105 @@ package main
...
+    if err != nil {
+      s.Error = err.Error()
+      return s
+    } else {
+      s.ExitCode = TaskIdCommandMap[task_id].Command.ProcessState.ExitCode()
```
Comment (a-palchikov) [inline execUtil.go]: The process needs to be marked as finished.

File: models.go [hunk 11]
```diff
@@ -0,0 +1,32 @@
+package main
+import (
+  "time"
+)
+
+// request and response structures
+
+// defines object returned to status handler
```
Comment (a-palchikov) [inline models.go]:
```go
// StatusResponse describes the status of a process
```
Also, document all exported attributes.

File: models.go [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+type StatusResponse struct {
+  Task_id string `json:"task_id"`
+  Finished bool `json:"finished"`
+  StartTime time.Time `json:"start_time"`
+  EndTime   time.Time `json:"end_time,omitempty"`
+  ExitCode  int `json:"exit_code,omitempty"`
```
Comment (a-palchikov) [inline models.go]: nit: Omitting `ExitCode` will mask the successful execution. This could instead be a pointer to int if you wanted to distinguish between set and unavailable exit code.

File: models.go [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+  Error   string `json:"error,omitempty"`
+}
+
+type StopResponse struct {
+  Task_id string `json:"task_id"`
+  ExitCode  int `json:"exit_code,omitempty"`
+  Error   string `json:"error,omitempty"`
+}
+
+type ErrorMessage struct {
```
Comment (a-palchikov) [inline models.go]: I have not seen this used anywhere.

File: models.go (hunk 12 above)
Comment (GeorgeLuo) [reply models.go]: I didn't notice this while testing, thanks

File: execUtil.go (hunk 9 above)
Comment (GeorgeLuo) [reply execUtil.go]: I am gathering that I should be running command /start with a process group and sending the interrupt signal to the group on /stop

File: models.go (hunk 12 above)
Comment (a-palchikov) [reply models.go]: No problem - it is a common issue. Note, that this will not be a problem in this case as unmarshaling the value will simply leave ExitCode uninitialized (and hence at `0`) but like I said there's no way of telling between set `0` and no exit code if one needed to distinguish.

File: execUtil.go (hunk 5 above)
Comment (GeorgeLuo) [reply execUtil.go]: I think I understand what you mean, go maps cannot perform read/write concurrently. I will implement with a sync.Map instance, if that is a valid solution. According to documentation, this is optimized for write mostly, which begs the question of purging records on /stop call. I think the logic for now will be to keep such records, which also is easier in interpretation when considering the /status calls on finished jobs (of which stopped processes are a subset).

File: execUtil.go (hunk 9 above)
Comment (a-palchikov) [reply execUtil.go]: Yes, but it's up to you if you want to invest into adding support for this. It is also fine if you just document this as a potential future enhancement.

File: execUtil.go (hunk 5 above)
Comment (GeorgeLuo) [reply execUtil.go]: What I have added is mutexes for the command map and command wrapper on get/set operations. I took the exit code and made it an attribute of the command wrapper, as this would otherwise cause data races due to the accessibility of the underlying command object (could also be resolved using go-cmd/cmd probably).

File: models.go (hunk 13 above)
Comment (GeorgeLuo) [reply models.go]: I have used this a few times in the handler on empty messages.

File: execUtil.go (hunk 4 above)
Comment (GeorgeLuo) [reply execUtil.go]: added checking.

File: execUtil.go [hunk 14]
```diff
@@ -1,62 +1,155 @@
...
+
+	task_id := hostname + "-" + strconv.Itoa(pgid) // TODO handle if Process or Pid nil
+	s.Task_id = task_id
+
+	taskIdCommandMap.Put(task_id, NewCommandWrapper(cmd, outBuf))
+
+	// async subroutine
+	go func() {
+		// TODO add append error to CommandWrapper, impl accessors and setters
+		err = cmd.Wait()
```
Comment (a-palchikov) [inline execUtil.go]: This will capture `err` from the calling func - there's no strict need to do that, it could be local to the goroutine.

File: execUtil.go [hunk 15]
```diff
@@ -1,62 +1,155 @@
...
+		s.Error = err.Error()
+		return s
+	}
+
+	task_id := hostname + "-" + strconv.Itoa(pgid) // TODO handle if Process or Pid nil
+	s.Task_id = task_id
+
+	taskIdCommandMap.Put(task_id, NewCommandWrapper(cmd, outBuf))
+
+	// async subroutine
```
Comment (a-palchikov) [inline execUtil.go]: Avoid comments that do not explain anything beyond the obvious - i.e. in this case, goroutines are async by definition.

File: execUtil.go [hunk 16]
```diff
@@ -1,62 +1,155 @@
...
+		}
+		cw.SetExitCode(cmd.ProcessState.ExitCode())
+		cw.SetFinished(true)
+	}()
+
+	return s
+}
+
+// StopProcess is called to end a previously started process.
+func StopProcess(task_id string) StopResponse {
```
Comment (a-palchikov) [inline execUtil.go]: Functions in this package take at least two responsibilities - the main responsibility of handling process start/query/stop loop and that of wrapping errors in some sort of response objects - apparently for the reason of being able to send them directly to client over the wire. This is sub-optimal - this package obviously has nothing to do with HTTP-based handling and should not hide errors inside a struct (at least errors that are not directly related to command), instead - return errors as expected and wrap them into responses/header codes in the package that implements HTTP layer.
This would mean returning error as usual:
```go
StopProcess(taskID string) (*StopResponse, error) {
```

File: execUtil.go [hunk 17]
```diff
@@ -1,62 +1,155 @@
...
+	var command *CommandWrapper
+	var ok bool
+
+	// if task_id invalid, return error.
+	if command, ok = taskIdCommandMap.Get(task_id); !ok {
+		s.Error = []string{"invalid task_id"}
+		return s
+	}
+
+	s.ExitCode = new(int)
```
Comment (a-palchikov) [inline execUtil.go]: nit: it is not going to work for all exit points - as some will leave it uninitialized which will have the same effect as before (i.e. it would appear that the process finished with `0` when it exited with an error).
you could return exit code as a pointer on-demand:
```go
s.ExitCode = intPtr(command.ExitCode())
// where, intPtr is something like
func intPtr(int value) *int {
	return &value
}
```

File: execUtil.go [hunk 18]
```diff
@@ -1,62 +1,155 @@
...
+	if command.GetFinished() {
+		s.Error = []string{"process already finished."}
+		*s.ExitCode = command.GetExitCode()
+		return s
+	}
+
+	// send stop signal first
+	// TODO verify flat processes work with gpid as well
+
+	pgid, err := syscall.Getpgid(command.Command.Process.Pid)
```
Comment (a-palchikov) [inline execUtil.go]: err is not handled

File: execUtil.go [hunk 19]
```diff
@@ -1,62 +1,155 @@
...
+	err = syscall.Kill(-pgid, syscall.SIGINT)
```
Comment (a-palchikov) [inline execUtil.go]: nit: Interrupt is not the ideal choice for a stop signal, I'd go with `SIGQUIT` or `SIGTERM`

File: execUtil.go [hunk 20]
```diff
@@ -1,62 +1,155 @@
...
+
+	if err != nil {
```
Comment (a-palchikov) [inline execUtil.go]: nit: if this is permission issue - there's no point retrying with `SIGKILL`

File: handlers.go [hunk 21]
```diff
@@ -1,85 +1,67 @@
...
 
 	vars := mux.Vars(r)
 	task_id := vars["task_id"]
 
-	status := GetProcessStatus(task_id)
-	log.Printf("task_id=%s, status=%d", task_id, status)
+	res := GetProcessStatus(task_id)
 
-	responseBody := StatusResponse{task_id, status}
 	w.WriteHeader(200)
```
Comment (a-palchikov) [inline handlers.go]: This should not default to `http.StatusOK` (also use status codes defined the by package instead of raw numeric value as these are hard to read unless memorized) and handle different cases of errors (like an invalid `taskID` which should return `http.StatusNotFound`)

File: handlers.go [hunk 22]
```diff
@@ -1,85 +1,67 @@
...
 
 	body, _ := ioutil.ReadAll(r.Body)
-	log.Printf("body=%s", body)
 
 	body_map := make(map[string]string)
 	json.Unmarshal(body, &body_map)
 
 	command := body_map["command"]
+	if command == "" {
+		w.WriteHeader(401)
```
Comment (a-palchikov) [inline handlers.go]: I wonder why this is returning unauthorized while something like `StatusBadRequest` would be more suitable?

File: handlers.go [hunk 23]
```diff
@@ -1,85 +1,67 @@
...
-	output := StopProcess(task_id)
+	if task_id == "" {
+		w.WriteHeader(401)
+		w.Header().Set("Content-Type", "application/json")
+		json.NewEncoder(w).Encode(ErrorMessage{"no task_id provided"})
+		return
+	}
 
-	log.Printf("task_id=%s, status=%s", task_id, output)
+	StopResponse := StopProcess(task_id)
```
Comment (a-palchikov) [inline handlers.go]: nit: use names consistently. Here, you name the response variable `StopResponse` while below one is called `res`.

File: main.go [hunk 24]
```diff
@@ -3,11 +3,11 @@ package main
 import (
 	"crypto/tls"
 	"crypto/x509"
```
Comment (a-palchikov) [inline main.go:5]: It would also be ideal if you published the code that generates the certificates so the program can be tested.

File: models.go [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"time"
+)
+
+// request and response structures
+
+// Response structure for /status call.
+type StatusResponse struct {
+	Task_id   string     `json:"task_id"`
```
Comment (a-palchikov) [inline models.go]: nit: Consistent naming, `TaskID` instead of `Task_id`.

File: models.go [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+	Finished  *bool      `json:"finished,omitempty"`
+	StartTime *time.Time `json:"start_time,omitempty"`
+	EndTime   *time.Time `json:"end_time,omitempty"`
+	ExitCode  *int       `json:"exit_code,omitempty"`
+	Output    []string   `json:"output,omitempty"`
+	Errors    []string   `json:"errors,omitempty"`
```
Comment (a-palchikov) [inline models.go]: It would be helpful if these attributes were documented.
Also, why errors as a slice not a a single error - where is it set, what kind of errors are these and how is client supposed to handle them?

File: syncOutput.go [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+	return rw.buf.Write(p)
+}
+
+// Lines returns the contents of the buffer at the current state.
+func (rw *Output) Lines() []string {
+	rw.Lock()
+	defer rw.Unlock()
+	s := bufio.NewScanner(rw.buf)
+	for s.Scan() {
+		rw.lines = append(rw.lines, s.Text())
```
Comment (a-palchikov) [inline syncOutput.go]: Since rw.lines are only used in this method - why make it an attribute? It could be a local variable.

**Review by a-palchikov — COMMENTED** [review]
It would be nice to see a couple of tests.

File: execUtil.go (hunk 17 above)
Comment (GeorgeLuo) [reply execUtil.go]: thank you, very helpful

File: execUtil.go (hunk 16 above)
Comment (GeorgeLuo) [reply execUtil.go]: yes, I was thinking along these lines but was trying to reduce the number of models. your suggestion makes sense.

File: main.go (hunk 24 above)
Comment (GeorgeLuo) [reply main.go:5]: perhaps I should do this in another pr? It would be relatively unproblematic in terms of merge conflicts.

File: models.go (hunk 25 above)
Comment (GeorgeLuo) [reply models.go]: just installed a plugin to that suggested this, implemented.

File: execUtil.go (hunk 20 above)
Comment (GeorgeLuo) [reply execUtil.go]: replacing with

`	if err != nil && err != syscall.EPERM {
`

File: models.go (hunk 26 above)
Comment (GeorgeLuo) [reply models.go]: I changed it back to a *string. I realized if the original call returned with an error, the StartCommand handler would not have executed the subroutine.

File: main.go (hunk 24 above)
Comment (GeorgeLuo) [reply main.go:5]: I included a file, generate-keys.sh to generate a self-signed certificate. The key and cert need to be delivered to the client, for now with the shared volume docker option it will work in testing though.

**Comment by GeorgeLuo** [comment]
important note in last commit, the status endpoint is a POST call now, for uniformity reasons and in the case batch status requests can more easily combine multiple task_id's.

File: client/client.go [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+
+	switch endpoint {
+	case "start":
+		request, err = StartRequest(body, host)
+	case "stop":
+		request, err = StopRequest(body, host)
+	case "status":
+		request, err = StatusRequest(body, host)
+	default:
+		err = errors.New("invalid command")
```
Comment (a-palchikov) [inline client/client.go:49]: `err` is not handled in any of these cases.

File: client/client.go [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+
+func main() {
+
+	if len(os.Args) < 2 {
+		fmt.Println("start, stop, or command instruction not provided")
+		os.Exit(1)
+	}
+
+	args := os.Args
+	var certFile string
```
Comment (a-palchikov) [inline client/client.go:28]: nit: it is always recommended to have flag defaults if possible as this greatly simplifies CLI. In this case, it would make sense to have defaults for cert/key files and default the `host` to `localhost`.
I'm not suggesting you implement that - just something to keep in mind.

File: client/client.go [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+	case "status":
+		request, err = StatusRequest(body, host)
+	default:
+		err = errors.New("invalid command")
+		os.Exit(1)
+	}
+
+	client, err = newClient(certFile, keyFile, certFile)
+	if err != nil {
+		fmt.Println(err.Error())
```
Comment (a-palchikov) [inline client/client.go:55]: nit: instead of repetitively handling the error I'd move the functionality to another func to simply return an error. Then in main, I'd only have to handle it once

File: client/client.go [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+	r, err := client.Do(request)
+	if err != nil {
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+
+	defer r.Body.Close()
+	responseBody, responseError := ioutil.ReadAll(r.Body)
+	if responseError != nil {
+		fmt.Println(err.Error())
```
Comment (a-palchikov) [inline client/client.go:68]: wrong `err` output

File: client/client.go [hunk 32]
```diff
@@ -0,0 +1,135 @@
...
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/GeorgeLuo/grpc/models"
+)
+
+// TODO persistent connection
+
+var client *http.Client
```
Comment (a-palchikov) [inline client/client.go:18]: I don't really see a need for the client being a local "global" variable - it seems to be only used inside `main`.

File: client/client.go [hunk 33]
```diff
@@ -0,0 +1,135 @@
...
+				Certificates: []tls.Certificate{cert},
+			},
+		},
+	}, nil
+}
+
+// TODO could return a config object holding this information
+
+// ReadArgs populates the common parameters between the 3 endpoints and returns request interface.
+func ReadArgs(args []string, endpoint *string, certFile *string, keyFile *string, host *string) (body interface{}, err error) {
```
Comment (a-palchikov) [inline client/client.go:104]: Honestly, I don't see how this method is helping evaluate the command line. I'd simply move it in-place to `main`.

Also adding sub-commands (`start`, `stop` and `status`) with corresponding sets of flags is probably a better way to gdo. This way you would not need conditionally support some flags.
And you don't have to make `body` opaque - i.e. no need to use `interface{}`

File: client/requests.go [hunk 34]
```diff
@@ -0,0 +1,47 @@
...
+	"encoding/json"
+	"errors"
+	"net/http"
+)
+
+// TODO consider refactor of these 3 methods. At the moment this makes sense,
+// but custom logic might be needed in the future.
+
+// StartRequest forms a request object for http to consume.
+func StartRequest(body interface{}, host string) (*http.Request, error) {
```
Comment (a-palchikov) [inline client/requests.go:14]: I'm wondering why you're using `interface{}` when you can accept a specific type (`StartRequest`)? This is inviting errors only visible at run-time when the compiler cannot type-check the parameters. This should be avoided unless absolutely necessary (not in this case, imho).

File: client/requests.go [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+	urlString := "https://" + host + ":8443/start"
+	byteBody, err := json.Marshal(body)
+	if err != nil {
+		return nil, errors.New("malformed body")
+	}
+	request, err := http.NewRequest("POST", urlString, bytes.NewBuffer(byteBody))
```
Comment (a-palchikov) [inline client/requests.go:20]: nit: `bytes.NewReader` would be sufficient here

File: client/requests.go [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+	request.Header.Set("Content-Type", "application/json")
+	return request, err
+}
+
+// StopRequest forms a request object for http to consume.
```
Comment (a-palchikov) [inline client/requests.go:25]: nit: this is a very generic comment being almost unhelpful. It should mention that the request object being built is a request to stop a (running) process.

File: client/requests.go [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+		return nil, errors.New("malformed body")
+	}
+	request, err := http.NewRequest("POST", urlString, bytes.NewBuffer(byteBody))
+	request.Header.Set("Content-Type", "application/json")
+	return request, err
+}
+
+// StatusRequest forms a request object for http to consume.
+func StatusRequest(body interface{}, host string) (*http.Request, error) {
+	urlString := "https://" + host + ":8443/status"
```
Comment (a-palchikov) [inline client/requests.go:39]: nit: all XYZRequest methods are the same except for the endpoint and could be implemented as:
```go
func StartRequest(req StartRequest, host string) (*http.Request, error) {
	return buildRequest(req, statusEndpoint(host))
}
```

File: execUtil.go [hunk 38]
```diff
@@ -1,62 +1,150 @@
...
-// start process
-func RunCommand(command string) string {
-  log.Printf("command=%s", command)
-  cmd := exec.Command(command)
-  cmd.Stdout = os.Stdout
-  err := cmd.Start()
-  if err != nil {
-     log.Fatal(err)
-  }
+// GetProcessStatus is called to retrieve the details of a processes by task_id.
```
Comment (a-palchikov) [inline execUtil.go:33]: nit: `// GetProcessStatus retrieves the status of the process specified with taskID`

File: execUtil.go [hunk 39]
```diff
@@ -1,62 +1,150 @@
...
+func GetProcessStatus(taskID string) (*models.StatusResponse, error) {
+
+	var statusResponse models.StatusResponse
+	statusResponse.TaskID = taskID
```
Comment (a-palchikov) [inline execUtil.go:37]: This is duplicated below and should not be here prior to validation anyways.

File: execUtil.go [hunk 40]
```diff
@@ -1,62 +1,150 @@
...
+	if command, ok = taskIDCommandMap.Get(taskID); !ok {
+		return nil, errors.New("invalid task_id")
+	}
+
+	// TODO refactor this into un/finished process
+	statusResponse.TaskID = taskID
+	statusResponse.StartTime = new(time.Time)
+	*statusResponse.StartTime = command.StartTime
+	statusResponse.ExecError = command.GetExecError()
+	statusResponse.Finished = new(bool)
```
Comment (a-palchikov) [inline execUtil.go:52]: nit: Like I mentioned before you don't really need this field as the same information is available with `statusResponse.EndTime != nil`. It could be hidden behind an `IsFinished()` method though.

File: handlers.go [hunk 41]
```diff
@@ -1,85 +1,93 @@
...
-	task_id := body_map["task_id"]
-	log.Printf("task_id=%s", task_id)
+	taskID := bodyMap["task_id"]
 
-	// TODO define and handle errors
-	output := StopProcess(task_id)
+	if taskID == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Header().Set("Content-Type", "application/json")
+		json.NewEncoder(w).Encode(models.ErrorMessage{TaskID: nil, Error: "no task_id provided"})
```
Comment (a-palchikov) [inline handlers.go:46]: nit: no need to specify `TaskID: nil` - it is the default.

File: handlers.go [hunk 42]
```diff
@@ -1,85 +1,93 @@
...
-	log.Printf("command=%s, task_id=%s", command, task_id)
 
-	responseBody := StartResponse{task_id}
-	w.WriteHeader(200)
+	bodyMap := make(map[string]string)
+	json.Unmarshal(body, &bodyMap)
+
+	command := bodyMap["command"]
+	if command == "" {
+		w.WriteHeader(http.StatusBadRequest)
```
Comment (a-palchikov) [inline handlers.go:74]: nit: this could be abstracted as a func:
```go
replyWithError(w, http.StatusBadRequest, "no command provided")
return
```

File: execUtil.go [hunk 43]
```diff
@@ -1,62 +1,150 @@
...
+		return &stopResponse, errors.New("process already finished")
+	}
+
+	// send stop signal first
+	// TODO verify flat processes work with gpid as well
+
+	pgid, _ := syscall.Getpgid(command.Command.Process.Pid)
+	err := syscall.Kill(-pgid, syscall.SIGQUIT)
+
+	if err != nil && err != syscall.EPERM {
```
Comment (a-palchikov) [inline execUtil.go:139]: It would be cool if stop command persisted the process's output.

File: models/models.go [hunk 44]
```diff
@@ -0,0 +1,55 @@
+package models
+
+import (
+	"time"
+)
+
+// request and response structures
+
+// StatusResponse is a structure to capture return for /status call.
+// Finished indicates the completion of a process (through termination or end of process with exit code).
```
Comment (a-palchikov) [inline models/models.go:10]: Field-specific doc-comments should be close to fields themselves:
```go
// StatusResponse describes the output of a /status call.
type StatusResponse struct {
	// TaskID identifies the process
	TaskID    string     `json:"task_id"`
```

File: syncMap.go [hunk 45]
```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"os/exec"
+	"sync"
+	"time"
+)
+
+// SyncMap is a map of taskID to previous processes.
```
Comment (a-palchikov) [inline syncMap.go:9]: You should mention that the type provides thread-safety to provide reasons for its existence.

File: syncMap.go [hunk 46]
```diff
@@ -0,0 +1,102 @@
...
+		StartTime:  time.Now(),
+		endTime:    nil,
+		StdoutBuff: outBuff,
+		exitCode:   cmd.ProcessState.ExitCode(),
+		execError:  nil,
+		Mutex:      &sync.Mutex{},
+	}
+}
+
+// SetEndTime is used to set the end time of a process.
```
Comment (a-palchikov) [inline syncMap.go:62]: `// SetEndTime sets the time when the process has finished execution`

File: syncMap.go [hunk 47]
```diff
@@ -0,0 +1,102 @@
...
+	return cmd, ok
+}
+
+// CommandWrapper encapsulates an exec.Cmd object with status metadata.
+type CommandWrapper struct {
+	Command    *exec.Cmd // underlying command
+	StartTime  time.Time
+	endTime    *time.Time
+	StdoutBuff *Output
+	execError  *string
```
Comment (a-palchikov) [inline syncMap.go:44]: Not sure why you'd use a pointer for string - I think a zero-value is usable here.

File: syncMap.go [hunk 48]
```diff
@@ -0,0 +1,102 @@
...
+// NewCommandWrapper is used to return a default CommandWrapper object.
+func NewCommandWrapper(cmd *exec.Cmd, outBuff *Output) *CommandWrapper {
+	return &CommandWrapper{
+		Command:    cmd,
+		StartTime:  time.Now(),
+		endTime:    nil,
+		StdoutBuff: outBuff,
+		exitCode:   cmd.ProcessState.ExitCode(),
+		execError:  nil,
+		Mutex:      &sync.Mutex{},
```
Comment (a-palchikov) [inline syncMap.go:58]: nit: While it's nice to be able to directly invoke methods on Mutex, exposing it does not have a benefit and could be used outside the scope of methods that rely on it, I'd hide the field.

File: test_process.sh [hunk 49]
```diff
@@ -1,7 +1,9 @@
 #!/bin/sh
+(
 sleep 1
 echo "I'm the script with pid $$"
-for i in 1 2 3 4 5 6 7 8 9 10; do
+for i in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20; do
```
Comment (a-palchikov) [inline test_process.sh:5]: `for i in $(seq 1 20); do`

File: handlers_test.go [hunk 50]
```diff
@@ -0,0 +1,189 @@
...
+	if !*statusResponse.Finished {
+		t.Errorf("command finished returned unexpected value: got %t want %t",
+			*statusResponse.Finished, true)
+	}
+}
+
+// TestStopProcessBasic validates stop command with status check after start then stop call.
+func TestStopProcessBasic(t *testing.T) {
+
+	var data = []byte(`{"command":"sleep 5"}`)
```
Comment (a-palchikov) [inline handlers_test.go:135]: The test cases are somewhat verbose and hard to follow. A test case should ideally only display the tested properties with all the configuration/setup abstracted away to minimize distraction:
```go
func TestStopProcessBasic(t *testing.T) {
	// setup
	// exercise
	var startResp StartResponse
	status := startProcess(t, "sleep 5", &startResp) // hides all of the logic used to start a process with testing

	// validate
	if status != http.StatusOK {
		t.Errorf("handler returned wrong status code: got %v want %v",
			status, http.StatusOK)
	}

	// exercise
	var stopResp StopResponse
	status := stopProcess(t, resp.TaskID, &stopResp)

	// validate status and stopResp.ExitCode
}
``` 
This follows the [four-phase](http://xunitpatterns.com/Four%20Phase%20Test.html) test strategy with setup/teardown surpressed.

**Review by a-palchikov — APPROVED** [review]
You should check in probably and can address the issues after if you have time for them, to minimize the PR scope.

File: client/client.go (hunk 29 above)
Comment (GeorgeLuo) [reply client/client.go:28]: completely agree. just did not get to yet

File: execUtil.go (hunk 43 above)
Comment (GeorgeLuo) [reply execUtil.go:139]: Can I get some clarification on this point? 

Are you talking about in the case the first SIGQUIT is sent and the error is overwritten with the subsequent SIGKILL? In that case I can add a field to CommandWrapper for quit_error. Or I can return an array of errors from the StopProcess handler.

File: client/client.go (hunk 33 above)
Comment (GeorgeLuo) [reply client/client.go:104]: I see. I wished not to repeat the cert and key flags for each of the subcommands so that's why I didn't define all the flagsets. This probably was reductionism at the expense of flexibility.

File: client/requests.go (hunk 37 above)
Comment (GeorgeLuo) [reply client/requests.go:39]: would it be appropriate to have buildRequest's req parameter be an interface at this point? Or defining some kind of inheritance structure for a Request? Or buildRequest could take a []byte as the body?

Comment (GeorgeLuo) [reply client/requests.go:39]: I think interface may be more reasonable here as buildRequest is internal to the requests.go context.

File: execUtil.go (hunk 43 above)
Comment (GeorgeLuo) [reply execUtil.go:139]: I think I understand what you're saying, syscall.Kill is it's own process with an output (?)

Comment (a-palchikov) [reply execUtil.go:139]: No, what I actually meant was that the process's output seemed to not be persisted after the stop request, so inspecting the stopped process (w/o fetching status while it was running) would return empty output. I was not sure this was the intended behavior.

Comment (GeorgeLuo) [reply execUtil.go:139]: MacBook-Pro-7:grpc mymac$ go run client/*go start -command ./test_process.sh 
{"task_id":"987f769fca40-39313"}

MacBook-Pro-7:grpc mymac$ go run client/*go stop -task_id 987f769fca40-39313
{"task_id":"987f769fca40-39313","exit_code":-1}

MacBook-Pro-7:grpc mymac$ go run client/*go status -task_id 987f769fca40-39313
{"task_id":"987f769fca40-39313","start_time":"2019-09-03T23:44:22.0981362Z","end_time":"2019-09-03T23:44:30.0391807Z","exit_code":-1,"output":["I'm the script with pid 39313","Still running 39313","Still running 39313","Still running 39313","Still running 39313","Still running 39313","Still running 39313"],"execError":"signal: quit"}

I see what you're talking about. The status command actually queries the output directly from the CommandWrapper (Command) object, so the output does persist. Now I could change that and add a field to the CommandWrapper to set output on os.Wait completion and stop command. But at the moment, the output will still exist.

Comment (GeorgeLuo) [reply execUtil.go:139]: Can't believe I've just noticed, retrieving a status consecutively eats the output the first time because it's from a buffer. I'll have to do as you suggested.
